go/ql/test/library-tests/semmle/go/aliases/DataflowFields/embedding.go

```diff
@@ -0,0 +1,29 @@
+package test
+
+import (
+	"test.com/basename/pkg1"
+	"test.com/basename/pkg2"
+)
+
+// Tests that dataflow behaves as expected when loads and stores from a field traverse embeddings that use different type names.
+
+// pkg2.IntStruct is an alias for pkg1.IntStruct
+// Note trickery with packages is necessary so that Go will assign the fields the same name as well as the same type.
+
+type EmbedsPkg1IntStruct = struct{ pkg1.IntStruct }
+type EmbedsPkg2IntStruct = struct{ pkg2.IntStruct }
```

Comment:
I think this part probably warrants a bit more explanation. What would happen without the package trickery? The comment implies that they will have different names and different types, but it's not immediately clear why that is or how it affects the test.

Reply:
This seems like a very artificial test, since I'm not sure why a package would define an alias without changing the name (just to avoid mentioning the original package?). More likely might be two packages with aliases with the same name for the same type. Would that test the same thing? Or be a separate test worth adding?

Reply:
In practice this happens because an impl package of some sort defines a type, then that type gets exposed, so you get something like

```
package impl
...
type IntStruct struct { x int }

type EmbedsIntStruct = struct { IntStruct }

...

package public
...
type IntStruct = impl.IntStruct

type EmbedsIntStruct = struct { IntStruct }
```

The promotion of the impl type to a public one makes sense. In the examples I saw I didn't see a particular reason to redefine the same type alias involving embedding twice in two different packages; probably just an accident that nobody had happened to rationalise. But the result was that there were two identical types that used different surface types for an embedded field, and we need to make sure our dataflow lib knows they are identical, so writes through one should connect with reads through the other. The specific change needed was in `hasOwnField`, which synthesises the implicit field's name.